frontend/src/components/charts/SubmissionsHeatmap.tsx

```diff
@@ -1,20 +1,60 @@
+import { useState } from "react";
 import { useTheme } from "@mui/material/styles/";
-import * as echarts from "echarts";
+import { format } from "date-fns";
 import ReactEcharts from "echarts-for-react";
 
-export const SubmissionsHeatmap = () => {
-  function getVirtualData(year: string) {
-    const date = +echarts.number.parseDate(year + "-01-01");
-    const end = +echarts.number.parseDate(+year + 1 + "-01-01");
-    const dayTime = 3600 * 24 * 1000;
+type SubmissionHeatmapProps = {
+  heatmapData:
+    | {
+        date: Date | string;
+      }[]
+    | undefined;
+};
+
+export const SubmissionsHeatmap = ({ heatmapData }: SubmissionHeatmapProps) => {
+  const theme = useTheme();
+  const [maxSubmissions, setMaxSubmissions] = useState<number>(0);
+
+  const processData = (
+    heatmapData:
+      | {
+          date: Date | string;
+        }[]
+      | undefined,
+  ) => {
+    const dateSubmissionMap = new Map<string, number>();
     const data = [];
-    for (let time = date; time < end; time += dayTime) {
-      // Data element consist of a date and a submission count
-      data.push([new Date(time), Math.floor(Math.random() * 1000)]);
+    if (!heatmapData) {
+      return;
+    } else {
+      heatmapData.map((data) => {
+        const formattedDate = format(new Date(data.date), "yyyy/M/d");
+        if (dateSubmissionMap.has(formattedDate)) {
+          const submissionCount = dateSubmissionMap.get(formattedDate);
+          if (!submissionCount) {
+            return;
+          } else {
+            const newSubmissionCount = submissionCount + 1;
+            dateSubmissionMap.set(formattedDate, newSubmissionCount);
+          }
+        } else {
+          dateSubmissionMap.set(formattedDate, 1);
+        }
+      });
+      for (const entry of dateSubmissionMap.entries()) {
+        const date = entry[0];
+        const submissions = entry[1];
+        data.push([new Date(date), submissions]);
+
+        // Find largest number of submissions in the year to set range of submission values
+        if (submissions > maxSubmissions) {
+          setMaxSubmissions(submissions);
+        }
+      }
     }
     return data;
-  }
-  const theme = useTheme();
+  };
+
```

Comment:
Not exactly sure what type is params in "formatter: (params: any)"

Reply:
According to apache echarts https://echarts.apache.org/en/option.html#tooltip.formatter its the same as the raw data that we give to the chart, so it should be `params: { data: [Date, number] }`.

Reply:
Also, for the formatter part (can't leave a comment there): Use `datefns` to format. And we should use `dd MMM YYYY` instead of `MMM dd YYYY`.